File: main/analysis/3_automatic_picking.py
import timeit
import numpy as np
import matplotlib.pyplot as plt

from all_functions import readBinaryMatrix, readBinaryArray, irls

from scipy.signal import savgol_filter
from scipy.ndimage import median_filter

#--------------------------------------------------------------------
start = timeit.default_timer()

nt = 3001
traces = 100
nodes_all = 121 
shots_all = 10000

dt = 1e-3

fw = 11

for node in range(nodes_all): 

    rawPicks_all = readBinaryArray(shots_all, f"../../inputs/picks/rawPicks_shot_{node+1}_{shots_all}_samples.bin")
    output_picks = np.zeros(shots_all)

    for line in range(traces):

        window = slice(int(line*shots_all/traces),int((line+1)*shots_all/traces))

        rawPicks = rawPicks_all[window].copy()    

        rawPicks[int(fw/2):-int(fw/2)] = median_filter(rawPicks[int(fw/2):-int(fw/2)], fw)

        output_picks[window] = savgol_filter(rawPicks, fw, 2)     

    print(f"File ../../inputs/picks/obsData_{shots_all}_samples_shot_{node+1}.bin was written successfully.")
    output_picks.astype("float32", order = "F").tofile(f"../../inputs/picks/obsData_{shots_all}_samples_shot_{node+1}.bin")

stop = timeit.default_timer()
print('Run time: ', stop - start)  

# Quality control 

tlag = 0.15
tcut = 4.50

updated_nt = 4501

times = slice(int(tlag/dt), int((tlag+tcut)/dt))

tloc = np.linspace(0, updated_nt-1, 11, dtype = int)
tlab = np.around(np.linspace(0, updated_nt-1, 11) * dt, decimals = 1)

xloc = np.linspace(0, traces-1, 11, dtype = int)
xlab = np.linspace(0, traces, 11, dtype = int)

for node in range(nodes_all):
    seismic_all = readBinaryMatrix(nt, shots_all, f"../../inputs/seismograms/seismogram_{nt}x{shots_all}_shot_{node+1}.bin") 
    rawPicks_all = readBinaryArray(shots_all, f"../../inputs/picks/rawPicks_shot_{node+1}_{shots_all}_samples.bin")
    output_picks = readBinaryArray(shots_all, f"../../inputs/picks/obsData_{shots_all}_samples_shot_{node+1}.bin")

    seismic = seismic_all[times, :]    

    scale = 0.5 * np.std(seismic)

    plt.figure(1, figsize = (20, 5))
    plt.imshow(seismic, aspect = "auto", cmap = "Greys", vmin = -scale, vmax = scale)
    plt.plot(rawPicks_all/dt, "o")
    plt.plot(output_picks/dt, "o")
    plt.title(f"{node+1}")
    plt.xlim([0, shots_all])
    plt.tight_layout()
    plt.show()

File: main/analysis/3_automatic_picking.cpp
# include <cmath>
# include <chrono>
# include <string>
# include <fstream>
# include <iostream>
# include <algorithm>

# include "../../src/essentials/utils.hpp"

int main()
{
    auto utils = Utils();

    auto ti = std::chrono::system_clock::now();

    int nt = 3001;
    int traces = 100;
    int nodes_all = 441; 
    int shots_all = 10000;

    float dt = 1e-3f;
    float tlag = 0.10f;
    float tcut = 2.80f;

    int updated_nt = (int)(tcut/dt);

    float window = 0.080f;
    int iw = (int)(window/dt);

    int itime = (int)(tlag/dt);
    int ftime = (int)((tcut+tlag)/dt + 1);

    // loop of receiver gathers
    for (int node = 0; node < nodes_all; node++)
    {
        float * picks_all = new float[shots_all]();    
        float * seismic_all = utils.readBinaryFloat("../../inputs/seismograms/seismogram_" + std::to_string(nt) + "x" + std::to_string(shots_all) + "_shot_" + std::to_string(node+1) + ".bin", shots_all*nt);
        
        std::cout<<"Running node " + std::to_string(node+1) + " of " + std::to_string(nodes_all)<<std::endl;

        // loop of traces
        for (int trace = 0; trace < shots_all; trace++)
        {
            float * A = new float[updated_nt]();
            float * B = new float[updated_nt]();
            float * S = new float[updated_nt]();

            if (trace % (shots_all/10) == 0)
            {
                std::cout<<"    Running trace " + std::to_string(trace) + " of " + std::to_string(shots_all)<<std::endl;
            }

            // loop of time
            for (int time = iw; time < updated_nt - iw; time++)
            {
                for (int k = 0; k < iw; k++)
                {
                    A[time] += seismic_all[(itime+time-k) + trace*nt] + 1e-15f;
                    B[time] += seismic_all[(itime+time+k) + trace*nt] + 1e-15f;
                }
                
                S[time] = fabsf(B[time]/A[time] * (B[time] - A[time]));
                S[time] *= (B[time] + A[time]) * (B[time] - A[time]) / powf(iw + time, 2.0f);
            }

            float ampMax = *std::max_element(S, S+updated_nt);

            // Normalizing and picking the first break
            for (int k = 0; k < updated_nt; k++)
            {
                S[k] *= 1.0f / ampMax;
                
                if (S[k] > 1e-4f)
                {
                    picks_all[trace] = (k + iw + iw/2) * dt;            
                    break;
                }
            }    

            delete[] A;
            delete[] B;
            delete[] S;
        }

        utils.writeBinaryFloat("../../inputs/picks/rawPicks_shot_" + std::to_string(node+1) + "_" + std::to_string(shots_all) + "_samples.bin", picks_all, shots_all);

        delete[] picks_all;
        delete[] seismic_all;
    }

    auto tf = std::chrono::system_clock::now();

    std::chrono::duration<double> elapsed_seconds = tf - ti;

    std::cout << "\nRun time: " << elapsed_seconds.count() << " s\n";

    return 0;
}